tests/programs/custom/zalrsc/zalrsc.c

```diff
@@ -0,0 +1,216 @@
+
+int main(int argc, char *argv[])
+{
+  volatile uint32_t failures=0;
+
+  num_misaligned_LRW_trapped = 0;
+  num_misaligned_SCW_trapped = 0;
+  unexpected_irq_beh = 0;
+  trap_handler_beh = UNEXPECTED_IRQ_BEH;
+
+  failures += test_LR_SC_sequence_success();
+  failures += test_LR_SC_sequence_failure_address_mismatch();
+  failures += test_LR_SC_sequence_failure_additional_SC_access();
+  failures += test_LRW_misaligned_access();
+  failures += test_SCW_misaligned_access();
+  failures += unexpected_irq_beh;
+
+  if(failures == 0){
+    return EXIT_SUCCESS;
+
+  } else if (unexpected_irq_beh) {
+    printf("\n\nTest run fails due to unexpected traps.\n\n");
+    return EXIT_FAILURE;
+
+  } else {
+    return EXIT_FAILURE;
+  }
+}
+
+int test_LR_SC_sequence_success(void)
+{
+  printf("\n\nTest LR.W SC.W sequence constructed to succeed.\n\n");
+
+  volatile uint32_t mem_addr = 0;
+  volatile uint32_t is_failure = 0;
+  volatile uint32_t random = 0;
+
+  mem_addr = (uint32_t)&mem_addr;
+
+  __asm__ volatile("lr.w %0, (%1)"     : "=r" (random)     : "r" (mem_addr));
+  __asm__ volatile("sc.w %0, %1, (%2)" : "=r" (is_failure) : "r" (random), "r" (mem_addr));
+
+  if (is_failure) {
+    printf("Expected sequence to succeed. However, it did not.\n");
+    return 1;
+  }
+
+  return 0;
+}
+
+int test_LR_SC_sequence_failure_address_mismatch(void)
+{
+  printf("\n\nTest LR.W SC.W sequence constructed to fail due to SC.W address mismatch.\n\n");
+
+  volatile uint32_t mem_addr_LRW = 0;
+  volatile uint32_t mem_addr_SCW = 0;
+  volatile uint32_t is_failure = 0;
+  volatile uint32_t random = 0;
+
+  mem_addr_LRW = (uint32_t)&mem_addr_LRW;
+  mem_addr_SCW = (uint32_t)&mem_addr_SCW;
+
+  __asm__ volatile("lr.w %0, (%1)"     : "=r" (random)      : "r" (mem_addr_LRW));
+  __asm__ volatile("sc.w %0, %1, (%2)" : "=r" (is_failure)  : "r" (random), "r" (mem_addr_SCW));
+
+  if (!is_failure) {
+    printf("Expected sequence to fail. However, it did not.\n");
+    return 1;
+  }
+
+  return 0;
+}
+
+int test_LR_SC_sequence_failure_additional_SC_access(void)
+{
+  printf("\n\nTest LR.W SC.W sequence constructed to fail due to additional SC.W access, before SC.W access the address LR.W original reserved.\n\n");
+
+  volatile uint32_t mem_addr;
+  volatile uint32_t mem_addr_additional_SCW;
+  volatile uint32_t is_failure_additional_SCW;
+  volatile uint32_t is_failure;
+  volatile uint32_t random;
+
+  mem_addr                = (uint32_t)&mem_addr;
+  mem_addr_additional_SCW = (uint32_t)&mem_addr_additional_SCW;
+
+  __asm__ volatile("lr.w %0, (%1)"     : "=r" (random)                    : "r" (mem_addr));
+  __asm__ volatile("sc.w %0, %1, (%2)" : "=r" (is_failure_additional_SCW) : "r" (random), "r" (mem_addr_additional_SCW)); //TODO: krdosvik, not 100% sure if this should fail?
+  __asm__ volatile("sc.w %0, %1, (%2)" : "=r" (is_failure)                : "r" (random), "r" (mem_addr));
```

Comment:
> //TODO: krdosvik, not 100% sure if this should fail?

Since there is no load reserved on `mem_addr_additional_SCW`, isn't that the reason it should fail?

Reply:
Thanks. No excuse, but I forgot where my TODOs was placed. But yeah, the spec states that sc.w invalidates all reservation made for a hart, and that was what I was couldnt remember, and was therefore unsure. But yeah, that is correct.